# Урок 6. Задача look-alike

In [1]:
from sklearn.metrics import recall_score, precision_score, roc_auc_score, accuracy_score, f1_score

def evaluate_results(y_test, y_predict):
    print('Classification results:')
    f1 = f1_score(y_test, y_predict)
    print("f1: %.2f%%" % (f1 * 100.0)) 
    roc = roc_auc_score(y_test, y_predict)
    print("roc: %.2f%%" % (roc * 100.0)) 
    rec = recall_score(y_test, y_predict, average='binary')
    print("recall: %.2f%%" % (rec * 100.0)) 
    prc = precision_score(y_test, y_predict, average='binary')
    print("precision: %.2f%%" % (prc * 100.0)) 
    return f1, roc, rec, prc

1. взять любой набор данных для бинарной классификации (можно скачать один из модельных с https://archive.ics.uci.edu/ml/datasets.php)

In [2]:
import pandas as pd
import numpy as np
df = pd.read_csv("./Data/arrhythmia.data", header=None)
df.head()

0    1    2    3    4    5    6    7    8    9    ...  270   271  272  273  \
0   75    0  190   80   91  193  371  174  121  -16  ...  0.0   9.0 -0.9  0.0   
1   56    1  165   64   81  174  401  149   39   25  ...  0.0   8.5  0.0  0.0   
2   54    0  172   95  138  163  386  185  102   96  ...  0.0   9.5 -2.4  0.0   
3   55    0  175   94  100  202  380  179  143   28  ...  0.0  12.2 -2.2  0.0   
4   75    0  190   80   88  181  360  177  103  -16  ...  0.0  13.1 -3.6  0.0   

   274  275  276   277   278  279  
0  0.0  0.9  2.9  23.3  49.4    8  
1  0.0  0.2  2.1  20.4  38.8    6  
2  0.0  0.3  3.4  12.3  49.0   10  
3  0.0  0.4  2.6  34.6  61.6    1  
4  0.0 -0.1  3.9  25.4  62.8    7  

[5 rows x 280 columns]

2. сделать feature engineering

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 452 entries, 0 to 451
Columns: 280 entries, 0 to 279
dtypes: float64(120), int64(155), object(5)
memory usage: 988.9+ KB


In [4]:
df.describe()

0           1           2           3           4           5    \
count  452.000000  452.000000  452.000000  452.000000  452.000000  452.000000   
mean    46.471239    0.550885  166.188053   68.170354   88.920354  155.152655   
std     16.466631    0.497955   37.170340   16.590803   15.364394   44.842283   
min      0.000000    0.000000  105.000000    6.000000   55.000000    0.000000   
25%     36.000000    0.000000  160.000000   59.000000   80.000000  142.000000   
50%     47.000000    1.000000  164.000000   68.000000   86.000000  157.000000   
75%     58.000000    1.000000  170.000000   79.000000   94.000000  175.000000   
max     83.000000    1.000000  780.000000  176.000000  188.000000  524.000000   

              6           7           8           9    ...         270  \
count  452.000000  452.000000  452.000000  452.000000  ...  452.000000   
mean   367.207965  169.949115   90.004425   33.676991  ...   -0.278982   
std     33.385421   35.633072   25.826643   45.431434  ...    0.548876   
min    232.000000  108.000000    0.000000 -172.000000  ...   -4.100000   
25%    350.000000  148.000000   79.000000    3.750000  ...   -0.425000   
50%    367.000000  162.000000   91.000000   40.000000  ...    0.000000   
75%    384.000000  179.000000  102.000000   66.000000  ...    0.000000   
max    509.000000  381.000000  205.000000  169.000000  ...    0.000000   

              271         272         273    274         275         276  \
count  452.000000  452.000000  452.000000  452.0  452.000000  452.000000   
mean     9.048009   -1.457301    0.003982    0.0    0.514823    1.222345   
std      3.472862    2.002430    0.050118    0.0    0.347531    1.426052   
min      0.000000  -28.600000    0.000000    0.0   -0.800000   -6.000000   
25%      6.600000   -2.100000    0.000000    0.0    0.400000    0.500000   
50%      8.800000   -1.100000    0.000000    0.0    0.500000    1.350000   
75%     11.200000    0.000000    0.000000    0.0    0.700000    2.100000   
max     23.600000    0.000000    0.800000    0.0    2.400000    6.000000   

              277         278         279  
count  452.000000  452.000000  452.000000  
mean    19.326106   29.473230    3.880531  
std     13.503922   18.493927    4.407097  
min    -44.200000  -38.600000    1.000000  
25%     11.450000   17.550000    1.000000  
50%     18.100000   27.900000    1.000000  
75%     25.825000   41.125000    6.000000  
max     88.800000  115.900000   16.000000  

[8 rows x 275 columns]

In [5]:
df[279].value_counts()

1     245
10     50
2      44
6      25
16     22
3      15
4      15
5      13
9       9
15      5
14      4
7       3
8       2
Name: 279, dtype: int64

Признак колонки 279 имеет 16 классов классификации (номер признака, название, количество):

01 - Normal 245  
02 - Ischemic changes (Coronary Artery Disease) 44  
03 - Old Anterior Myocardial Infarction 15  
04 - Old Inferior Myocardial Infarction 15  
05 - Sinus tachycardy 13  
06 - Sinus bradycardy 25  
07 - Ventricular Premature Contraction (PVC) 3  
08 - Supraventricular Premature Contraction 2  
09 - Left bundle branch block 9  
10 - Right bundle branch block 50  
11 - 1. degree AtrioVentricular block 0  
12 - 2. degree AV block 0  
13 - 3. degree AV block 0  
14 - Left ventricule hypertrophy 4  
15 - Atrial Fibrillation or Flutter 5  
16 - Others 22

Буду классифицировать 06 - Sinus Bradycardy

In [6]:
df[279].replace([1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16], 0, inplace=True)
df[279].replace(6, 1, inplace=True)
df[279].value_counts()

0    427
1     25
Name: 279, dtype: int64

In [7]:
def get_num_features(df):
    """
    функция возвращает список числовых признаков
    :param df
    :return: list
    """
    numerical_features = df.select_dtypes(include=[np.number])
    # print(f"count of numeric_features_train {numerical_features.shape[1]}")
    return numerical_features.columns


def get_cat_features(df):
    """
    функция возвращает список признаков с типом "объект"
    :param df:
    :return: list
    """
    categorical_features = df.select_dtypes(include=[np.object])
    return categorical_features.columns

In [8]:
cat_list = get_cat_features(df)

/tmp/ipykernel_9791/4286977619.py:18: DeprecationWarning: `np.object` is a deprecated alias for the builtin `object`. To silence this warning, use `object` by itself. Doing this will not modify any behavior and is safe. 
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  categorical_features = df.select_dtypes(include=[np.object])


In [9]:
df[cat_list].value_counts()

10  11  12  13    14 
-1  46  -1  ?     85     1
47  48  35  ?     68     1
52  -4  0   ?     86     1
51  67  34  -124  77     1
    61  55  ?     61     1
                        ..
17  46  25  ?     79     1
    37  68  ?     123    1
    36  54  -118  64     1
    16  47  ?     57     1
?   64  -1  ?     85     1
Length: 452, dtype: int64

In [10]:
df[10].unique()

array(['13', '37', '34', '11', '66', '49', '7', '69', '71', '42', '51',
       '20', '45', '75', '-24', '28', '39', '78', '56', '10', '17', '112',
       '52', '48', '153', '172', '16', '32', '46', '23', '35', '3', '-8',
       '-5', '4', '60', '8', '-18', '76', '64', '9', '79', '33', '14',
       '30', '53', '25', '62', '?', '43', '31', '19', '50', '68', '-164',
       '-56', '15', '-2', '87', '-136', '175', '-124', '142', '-13', '-9',
       '58', '47', '-165', '73', '21', '38', '70', '-92', '143', '22',
       '72', '57', '41', '65', '36', '63', '55', '44', '77', '24', '81',
       '12', '29', '18', '40', '54', '-1', '104', '-167', '-15', '67',
       '108', '-75', '160', '114', '-22', '120', '-115', '147', '-82',
       '-63', '-172', '102', '179', '86', '-52', '-49', '174', '-27',
       '95', '-123', '61', '137', '164', '5', '-177', '-103', '-106',
       '-42', '117', '127', '-160', '122', '59', '-16', '26', '100', '0',
       '-21', '-121', '-174', '-51', '105', '-116', '-4', '

In [11]:
for col in cat_list:
    # извлекаем индексы по условию (тех элементов, которые имеют значение "?")
    indices = np.where(df[col] != '?')
    # получаем спиcок элементов по индексам [0], приводим к строке, к целому и находим медиану
    col_med = np.median(np.take(df[col], indices[0]).astype(str).astype(int))
    # заменяем некорректное значение медианой
    df[col] = df[col].replace('?', col_med)    

In [12]:
for col in cat_list:
    df[col] = df[col].astype(int)

In [13]:
df[cat_list].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 452 entries, 0 to 451
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   10      452 non-null    int64
 1   11      452 non-null    int64
 2   12      452 non-null    int64
 3   13      452 non-null    int64
 4   14      452 non-null    int64
dtypes: int64(5)
memory usage: 17.8 KB


3. обучить любой классификатор (какой вам нравится)

In [14]:
from sklearn.model_selection import train_test_split

x_data = df.iloc[:,:-1]
y_data = df.iloc[:,-1]

x_train, x_test, y_train, y_test = train_test_split(x_data, y_data, test_size=0.2, random_state=42)

In [15]:
import xgboost as xgb

model = xgb.XGBClassifier()

model.fit(x_train, y_train)
y_predict = model.predict(x_test)

/home/mrp/anaconda3/envs/ml_business/lib/python3.8/site-packages/xgboost/sklearn.py:1146: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[18:42:09] WARNING: /home/conda/feedstock_root/build_artifacts/xgboost-split_1631904754241/work/src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


In [16]:
f1, roc, rec, prc = evaluate_results(y_test, y_predict)

score = []
score.append([f1, prc, rec, roc])

Classification results:
f1: 50.00%
roc: 66.67%
recall: 33.33%
precision: 100.00%


4. далее разделить ваш набор данных на два множества: P (positives) и U (unlabeled). Причем брать нужно не все положительные (класс 1) примеры, а только лишь часть

In [17]:
mod_data = df.copy()
#get the indices of the positives samples
pos_ind = np.where(mod_data.iloc[:,-1].values == 1)[0]
#shuffle them
np.random.shuffle(pos_ind)
# leave just 25% of the positives marked
pos_sample_len = int(np.ceil(0.25 * len(pos_ind)))
print(f'Using {pos_sample_len}/{len(pos_ind)} as positives and unlabeling the rest')
pos_sample = pos_ind[:pos_sample_len]

Using 7/25 as positives and unlabeling the rest


In [18]:
mod_data['class_test'] = -1
mod_data.loc[pos_sample,'class_test'] = 1
print('target variable:\n', mod_data.iloc[:,-1].value_counts())

target variable:
 -1    445
 1      7
Name: class_test, dtype: int64


5. применить random negative sampling для построения классификатора в новых условиях

In [19]:
mod_data = mod_data.sample(frac=1)
neg_sample = mod_data[mod_data['class_test']==-1][:len(mod_data[mod_data['class_test']==1])]
sample_test = mod_data[mod_data['class_test']==-1][len(mod_data[mod_data['class_test']==1]):]
pos_sample = mod_data[mod_data['class_test']==1]
print(neg_sample.shape, pos_sample.shape)
sample_train = pd.concat([neg_sample, pos_sample]).sample(frac=1)

(7, 281) (7, 281)


In [20]:
model = xgb.XGBClassifier()

model.fit(sample_train.iloc[:,:-2].values, 
          sample_train.iloc[:,-2].values)
y_predict = model.predict(sample_test.iloc[:,:-2].values)

f1, prc, rec, roc = evaluate_results(sample_test.iloc[:,-2].values, y_predict)

[18:42:10] WARNING: /home/conda/feedstock_root/build_artifacts/xgboost-split_1631904754241/work/src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


/home/mrp/anaconda3/envs/ml_business/lib/python3.8/site-packages/xgboost/sklearn.py:1146: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


Classification results:
f1: 13.42%
roc: 63.37%
recall: 55.56%
precision: 7.63%


In [21]:
score.append([f1, prc, rec, roc])

6. сравнить качество с решением из пункта 3 (построить отчет - таблицу метрик)

In [22]:
score_df = pd.DataFrame({'Пункт 3': score[0], 'Пункт 5': score[1]}).T
score_df.columns = ['fscore', 'precision', 'recall', 'roc_auc']
score_df

fscore  precision    recall   roc_auc
Пункт 3  0.500000    1.00000  0.333333  0.666667
Пункт 5  0.134228    0.63373  0.555556  0.076336

7. поэкспериментировать с долей P на шаге 5 (как будет меняться качество модели при уменьшении/увеличении размера P)

In [23]:
mod_data = df.copy()
#get the indices of the positives samples
pos_ind = np.where(mod_data.iloc[:,-1].values == 1)[0]
#shuffle them
np.random.shuffle(pos_ind)
# leave just 30% of the positives marked
pos_sample_len = int(np.ceil(0.3 * len(pos_ind)))
print(f'Using {pos_sample_len}/{len(pos_ind)} as positives and unlabeling the rest')
pos_sample = pos_ind[:pos_sample_len]

Using 8/25 as positives and unlabeling the rest


In [24]:
mod_data['class_test'] = -1
mod_data.loc[pos_sample,'class_test'] = 1
print('target variable:\n', mod_data.iloc[:,-1].value_counts())

target variable:
 -1    444
 1      8
Name: class_test, dtype: int64


In [25]:
mod_data = mod_data.sample(frac=1)
neg_sample = mod_data[mod_data['class_test']==-1][:len(mod_data[mod_data['class_test']==1])]
sample_test = mod_data[mod_data['class_test']==-1][len(mod_data[mod_data['class_test']==1]):]
pos_sample = mod_data[mod_data['class_test']==1]
print(neg_sample.shape, pos_sample.shape)
sample_train = pd.concat([neg_sample, pos_sample]).sample(frac=1)

(8, 281) (8, 281)


In [26]:
model = xgb.XGBClassifier()

model.fit(sample_train.iloc[:,:-2].values, 
          sample_train.iloc[:,-2].values)
y_predict = model.predict(sample_test.iloc[:,:-2].values)

f1, prc, rec, roc = evaluate_results(sample_test.iloc[:,-2].values, y_predict)

score.append([f1, prc, rec, roc])

[18:42:11] WARNING: /home/conda/feedstock_root/build_artifacts/xgboost-split_1631904754241/work/src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


/home/mrp/anaconda3/envs/ml_business/lib/python3.8/site-packages/xgboost/sklearn.py:1146: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


Classification results:
f1: 66.67%
roc: 95.21%
recall: 93.75%
precision: 51.72%


In [27]:
score_df = pd.DataFrame({'Пункт 3': score[0], 'Пункт 5': score[1], 'Пункт 7_0.3': score[2]}).T
score_df.columns = ['fscore', 'precision', 'recall', 'roc_auc']
score_df

fscore  precision    recall   roc_auc
Пункт 3      0.500000   1.000000  0.333333  0.666667
Пункт 5      0.134228   0.633730  0.555556  0.076336
Пункт 7_0.3  0.666667   0.952083  0.937500  0.517241

In [28]:
mod_data = df.copy()
#get the indices of the positives samples
pos_ind = np.where(mod_data.iloc[:,-1].values == 1)[0]
#shuffle them
np.random.shuffle(pos_ind)
# leave just 40% of the positives marked
pos_sample_len = int(np.ceil(0.4 * len(pos_ind)))
print(f'Using {pos_sample_len}/{len(pos_ind)} as positives and unlabeling the rest')
pos_sample = pos_ind[:pos_sample_len]

Using 10/25 as positives and unlabeling the rest


In [29]:
mod_data['class_test'] = -1
mod_data.loc[pos_sample,'class_test'] = 1
print('target variable:\n', mod_data.iloc[:,-1].value_counts())

target variable:
 -1    442
 1     10
Name: class_test, dtype: int64


In [30]:
mod_data = mod_data.sample(frac=1)
neg_sample = mod_data[mod_data['class_test']==-1][:len(mod_data[mod_data['class_test']==1])]
sample_test = mod_data[mod_data['class_test']==-1][len(mod_data[mod_data['class_test']==1]):]
pos_sample = mod_data[mod_data['class_test']==1]
print(neg_sample.shape, pos_sample.shape)
sample_train = pd.concat([neg_sample, pos_sample]).sample(frac=1)

(10, 281) (10, 281)


In [31]:
model = xgb.XGBClassifier()

model.fit(sample_train.iloc[:,:-2].values, 
          sample_train.iloc[:,-2].values)
y_predict = model.predict(sample_test.iloc[:,:-2].values)

f1, prc, rec, roc = evaluate_results(sample_test.iloc[:,-2].values, y_predict)

score.append([f1, prc, rec, roc])

[18:42:12] WARNING: /home/conda/feedstock_root/build_artifacts/xgboost-split_1631904754241/work/src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


/home/mrp/anaconda3/envs/ml_business/lib/python3.8/site-packages/xgboost/sklearn.py:1146: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


Classification results:
f1: 65.12%
roc: 94.99%
recall: 93.33%
precision: 50.00%


In [32]:
score_df = pd.DataFrame({'Пункт 3': score[0], 'Пункт 5': score[1], 'Пункт 7_0.3': score[2], 'Пункт 7_0.4': score[3]}).T
score_df.columns = ['fscore', 'precision', 'recall', 'roc_auc']
score_df

fscore  precision    recall   roc_auc
Пункт 3      0.500000   1.000000  0.333333  0.666667
Пункт 5      0.134228   0.633730  0.555556  0.076336
Пункт 7_0.3  0.666667   0.952083  0.937500  0.517241
Пункт 7_0.4  0.651163   0.949880  0.933333  0.500000